### PR TITLE
[Experiment] Add profiler @open sesame 12/08 15:02

### LINKED ISSUE #787: Add a timing function for experiment
Linked with issue tracked with https://github.sec.samsung.net/jhoon-it-lee/nntrainer-experiments/issues/7

### PR DESCRIPTION
**Pull request by zhoonit**
- ~**#785** [Meson] Add ndk-build to be part of ndk build~
- [Experiment] Add profiler
```
This patch add `enable-profile` option to enable profile. Also this
patch adds a simple profiling logic to `neuralnet::inference`

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```
cc. @kparichay 

resolves #787

### CODE REVIEW

File: nntrainer/models/neuralnet.h [hunk 1]
```diff
@@ -345,6 +345,18 @@ class NeuralNetwork : public ml::train::Model {
    */
   void printPreset(std::ostream &out, unsigned int preset);
 
+#ifdef PROFILE
+  /**
+   * @brief query profile result
+   *
+   * @return double profile result
+   */
+  long int profileResult() { return time_taken; };
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:354]: how about `uint64_t`?

File: nntrainer/models/neuralnet.h [hunk 2]
```diff
@@ -345,6 +345,18 @@ class NeuralNetwork : public ml::train::Model {
...
+
+private:
+  long int time_taken;
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:357]: add some comment here? 
and how about forward_time_taken? as there will be a backward timing variable later too.

**Review by kparichay — COMMENTED** [review]
Overall LGTM other than the minor comment

File: nntrainer/models/neuralnet.h [hunk 3]
```diff
@@ -345,6 +345,18 @@ class NeuralNetwork : public ml::train::Model {
    */
   void printPreset(std::ostream &out, unsigned int preset);
 
+#ifdef PROFILE
+  /**
+   * @brief query profile result
+   *
+   * @return double profile result
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:352]: is this double?

File: nntrainer/models/neuralnet.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: Done

File: nntrainer/models/neuralnet.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: Regarding this,  I encapsulated and modularized the profiler into an inner class. Please take a look. Thanks for the suggestion

File: nntrainer/models/neuralnet.h (hunk 3 above)
Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: Fixed

File: nntrainer/models/neuralnet.h [hunk 4]
```diff
@@ -354,6 +357,70 @@ class NeuralNetwork : public ml::train::Model {
     manager.setGradientMemoryOptimization(opt);
   }
 
+/// @todo Make a more common class have this
+/// Maybe appcontext can have this?
+#ifdef PROFILE
+  class Profiler {
+  public:
+    // Designated key.
+    enum { SINGLE_FORWARD = 0 };
```
Comment (kparichay) [inline nntrainer/models/neuralnet.h:366]: is this SINGLE_FORWARD intended to give time for a single forward. If yes, then maybe divide by count maybe needed. If not, then maybe the SINGLE prefix can be removed?

Comment (zhoonit) [reply nntrainer/models/neuralnet.h]: I think it is a bit misleading, changed.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**